src/input_common/gcadapter/gc_adapter.cpp

```diff
@@ -0,0 +1,371 @@
+// Copyright 2014 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <chrono>
+#include <thread>
+#include "common/logging/log.h"
+#include "input_common/gcadapter/gc_adapter.h"
+
+namespace GCAdapter {
+
+Adapter::Adapter() {
+    if (usb_adapter_handle != nullptr) {
+        return;
+    }
+    LOG_INFO(Input, "GC Adapter Initialization started");
+
+    current_status = NO_ADAPTER_DETECTED;
+    libusb_init(&libusb_ctx);
+
+    StartScanThread();
+}
+
+GCPadStatus Adapter::CheckStatus(int port, const std::array<u8, 37>& adapter_payload) {
+    GCPadStatus pad = {};
+    bool get_origin = false;
+
+    ControllerTypes type = ControllerTypes(adapter_payload[1 + (9 * port)] >> 4);
+    if (type != ControllerTypes::None)
+        get_origin = true;
+
+    adapter_controllers_status[port] = type;
+
+    if (adapter_controllers_status[port] != ControllerTypes::None) {
+        u8 b1 = adapter_payload[1 + (9 * port) + 1];
+        u8 b2 = adapter_payload[1 + (9 * port) + 2];
+
+        if (b1 & (1 << 0)) {
+            pad.button |= PAD_BUTTON_A;
+        }
+        if (b1 & (1 << 1)) {
+            pad.button |= PAD_BUTTON_B;
+        }
+        if (b1 & (1 << 2)) {
+            pad.button |= PAD_BUTTON_X;
+        }
+        if (b1 & (1 << 3)) {
+            pad.button |= PAD_BUTTON_Y;
+        }
+
+        if (b1 & (1 << 4)) {
+            pad.button |= PAD_BUTTON_LEFT;
+        }
+        if (b1 & (1 << 5)) {
+            pad.button |= PAD_BUTTON_RIGHT;
+        }
+        if (b1 & (1 << 6)) {
+            pad.button |= PAD_BUTTON_DOWN;
+        }
+        if (b1 & (1 << 7)) {
+            pad.button |= PAD_BUTTON_UP;
+        }
```

Comment:
This *might* be a little better (not sure if it breaks anything)
Once again formatting is broken

```suggestion
enum b1_pad_buttons : int {
PAD_BUTTON_A,
PAD_BUTTON_B,
PAD_BUTTON_X,
PAD_BUTTON_Y,
PAD_BUTTON_LEFT,
PAD_BUTTON_RIGHT,
PAD_BUTTON_DOWN,
PAD_BUTTON_UP
};

constexpr std::array<int, 8> b1_buttons {
PAD_BUTTON_A, PAD_BUTTON_B, PAD_BUTTON_X, PAD_BUTTON_Y,
PAD_BUTTON_LEFT, PAD_BUTTON_RIGHT, PAD_BUTTON_DOWN, PAD_BUTTON_UP
};

for(int j = 0; j < b1_buttons.size(); j++) {
        if (b1 & (1 << j)) {
            pad.button |= b1_buttons[j];
        }
}
```

Reply:
I was debating whether to keep these verbose, or to do a trick like you are proposing. gc_poller.cpp line 94 has a similar approach. I may be wrong but either approach probably won't have any performance difference. 

> 
> 
> This _might_ be a little better (not sure if it breaks anything)

I was debating whether to keep these verbose, or to do a trick like you are proposing. gc_poller.cpp line 94 has a similar approach. I may be wrong but either approach probably won't have any performance difference.

Reply:
@lioncash do you have a preference for which approach we should move forward with?

Reply:
I addressed this in the latest commit. LMK what you think.